### ISSUE TITLE
Switch from a three-thread configuration to Boost.Asio

### ISSUE THREAD

**Issue by erniejunior** [ZeroCM/zcm#260]
Right now the user as no choice but to spawn at least two threads for sending and for receiving and possibly a third handler thread (see issue #258). Switching to Boost.Asio would allow to use any number of threads and we could possibly get rid of most of the locks.

**Comment by jbendes** [comment]
What is the rationale here? Why do you perceive the way it is right now to be bad?

**Comment by erniejunior** [comment]
**Short answer:** 
Performance.

**Long answer:**
I started with some very simple performance tests and immediately got some weird results (see issue #261). In parallel I looked at the implementation of the blocking library and found the following interesting points:
* ZCM uses its own queue and concurrent queue implementation which involves lots of locks and other synchronization mechanisms. In issue #259 I propose using a queue that is written and tested by queue-experts and which comes without the need for any locking.
* Only a minor thing, but the receiver thread as well as the handle thread **both** look up the channel name of an incoming message in the subscription list. If we solve this smarter we can save us half of those lookups.
* Any ZCM application is forced to run at least three threads. Threads do mean some overhead on their own and the required synchronization mechanisms even more. 

Here I would like to cite [the readme of the concurrent queue that I proposed](https://github.com/cameron314/concurrentqueue#reasons-not-to-use):
> The fastest synchronization of all is the kind that never takes place.

And from the [Asio documentation](http://think-async.com/Asio/asio-1.12.2/doc/asio/overview/rationale.html)
> Asio provides the tools to manage [...] long running operations, without requiring programs to use concurrency models based on threads and explicit locking.

At the same time, it is scalable in the sense that it allows to use any number of worker threads as illustrated in [this article](https://theboostcpplibraries.com/boost.asio-io-services-and-io-objects).

I hope this makes my point more clear.

**Comment by jbendes** [comment]
I'm all game for the use of a better queue. I am not familiar enough with the concurrent queue to comment on its implementation. Aside from code quality and functionality, my top concern would be licensing and compatibility. The queue would have to work on any OS (I don't want to move farther from OS uniformity). It would have to not add any build dependencies (a key design principle of zcm is that it has _very_ limited required external dependencies). And last but not least it must have a license that will allow us to make it part of the repo which is covered under LGPL.

**Comment by jbendes** [comment]
I don't think I follow the suggestion here. Maybe a pseudocode example would help me understand? I am a huge fan of unifying the threaded nature (especially if it gets us OS uniformity), but please understand skepticism and hesitation with changing the underlying blocking zcm implementation. I'm all for change for the better (and don't let me hesitation stop you from submitting a pullreq if you're gung-ho about doing this), but I'd like to understand the end result of what you're suggesting a bit more

**Comment by erniejunior** [comment]
I understand your concerns also adding boost as a dependency is a big step!

I think it is hard to explain if you never used Asio or read about it. Using Asio would mean for ZCM that we could do everything we did before *without multithreading* and *without locking mechanisms*. This would potentially *simplify the code* a lot. In cases where (multi-threaded) ZCM uses less than an entire CPU core it could *speed things up* because we get rid of the threading and locking overhead. In the (hopefully rare) cases where ZCM is so busy that it would require more than an entire CPU core, it is easy to introduce multi-threading again using the methods described [here](https://theboostcpplibraries.com/boost.asio-io-services-and-io-objects)

Some resources on Asio:
* Talk by its author [video](https://www.youtube.com/watch?v=D-lTwGJRx0o), [slides](https://raw.github.com/boostcon/2011_presentations/master/mon/thinking_asynchronously.pdf)
* [What is Boost.Asio, and why we should use it](http://alexott.net/en/cpp/BoostAsioNotes.html)

**Comment by jbendes** [comment]
Yeah I'm not thrilled with adding boost as a *required* dependency. I also just still don't see the benefit. You would still need a separate thread to detach from the main user thread and at that point you still need all the same locking, right? Again not opposed if there's a large benefit, I just don't see it yet.

**Comment by erniejunior** [comment]
I understand your concerns about adding the boost dependency. Maybe just switching to a different queue and doing some other small fixes is already enough to gain some performance.

Like with ZCM, where you can decide to call `start()` to spawn another thread or `run()` to run the handlers in the current thread Asio does give you the option to run everything in one thread.

**Comment by jbendes** [comment]
ahhhh i see now. I mean whipping up a parallel version of blocking.cpp using asio isn't crazy at all. We could have it be enabled with a --use-asio if you'd like.
